client/system_key.go

```diff
@@ -0,0 +1,127 @@
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package client
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/url"
+	"time"
+)
+
+type SystemKeyMismatchAction string
+
+const (
+	MismatchActionProceed       SystemKeyMismatchAction = "proceed"
+	MismatchActionWaitForChange SystemKeyMismatchAction = "wait-for-change"
+)
+
+var (
+	// ErrMismatchAdviceUnsupported indicates that current snapd does not
+	// support the API to obtain advice on system key mismatch.
+	ErrMismatchAdviceUnsupported = errors.New("system-key mismatch action is not supported")
+)
+
+type AdvisedAction struct {
+	SuggestedAction SystemKeyMismatchAction
+	// ChangeID carries the change to wait for if action was "wait-for-change".
+	ChangeID string
+}
+
+// SystemKeyMismatchAdvise is usually called after having detected a mismatch of
+// the system keys and performs an access to snapd API with the goal of getting
+// an advice from snapd on how to proceed. The request carries system-key
+// derived by the client. Returns an error of type *Error for errors
+// communicated directly through snapd API. Note, that the response is only
+// advisory.
+func (client *Client) SystemKeyMismatchAdvise(systemKey string) (*AdvisedAction, error) {
+	if systemKey == "" {
+		return nil, fmt.Errorf("no system key provided")
+	}
+
+	// same as for /v2/system-info
+	opts := &doOptions{
+		Timeout: 25 * time.Second,
+		Retry:   doRetry,
+	}
+
+	// update maintenance status
+	client.checkMaintenanceJSON()
+
+	var rsp response
+	var body io.Reader
+
+	q := url.Values{}
+	q.Add("system-key-mismatch", "true")
+
+	d, doErr := json.Marshal(struct {
+		SystemKey string `json:"system-key"`
+	}{
+		SystemKey: systemKey,
+	})
+	if doErr != nil {
+		return nil, fmt.Errorf("cannot marshal request data: %w", doErr)
+	}
+
+	body = bytes.NewReader(d)
+
+	statusCode, doErr := client.do("POST", "/v2/system-info", q, nil, body, &rsp, opts)
+	if statusCode == 405 {
+		// Method Not Allowed, most likely we are talking to an older version of
+		// snapd, which does not support the advisory system key mismatch
+		// handling, but since we got a response, snapd must be up already
+		return nil, ErrMismatchAdviceUnsupported
+	}
+
+	if doErr != nil {
+		return nil, doErr
+	}
+
+	if err := rsp.err(client, statusCode); err != nil {
+		return nil, err
+	}
+
+	var act *AdvisedAction
+
+	switch rsp.Type {
+	case "sync":
+		// body in the request?
+		act = &AdvisedAction{
```

Comment:
With the control flow tweak I suggested this could just return the advised action here.